Add initial support NXP MCXN947

mmahadevan108 opened the pull request:
This PR adds initial support for UART console and GPIO for NXP MCXN947.

Commit:
drivers: gpio_mcux: Add support for GPIO ICR

Add support for devices that use the kinetis
GPIO/PORT scheme but with the slight modification
that the interrupt control bits are in the GPIO
peripheral instead of the PORT peripheral.

Signed-off-by: Declan Snyder <declan.snyder@nxp.com>

mmahadevan108 committed:
drivers: pinctrl: Add support for more ports

MCX has more port modules which requires updating
the pinctrl driver

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: mcxn94x: Add initial support for NXP MCXN94X

Add initial support for NXP MCXN94X

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: pinctrl: Add input buffer for kinetis pinctrl

Some versions that use the kinetis pinctrl use the
input buffer

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
dts: nxp: Add binding for NXP Flexcomm device

This binding is for the NXP FlexComm Interface. The
driver will setup an interface for use as a UART, SPI
or I2C device.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: mfd: Add NXP LP Flexcomm driver

1. This driver allows to configure the Low Power Flexcomm
   interface as a UART, SPI or I2C device.
2. Manages the interrupts and dispatches it to the
   appropriate device driver.
3. A Flexcomm interface can be configured to be used as
   I2C and UART device. However a SPI device cannot be used
   concurrently with I2C or UART. Add checks for this feature.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: clock: Enable PORT & LP Flexcomm in NXP syscon driver

1. Add support to enable PORT control clocks
2. Add support for LP Flexcomm
3. MCX family has a different Clock API for USDHC.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
soc: mcxn947: Add support for NXP MCXN947

Add initial support for NXP MCXN947 SoC

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>
Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
boards: frdm_mcxn947: Add support for frdm_mcxn947 board

Add support for frdm_mcxn947 board

Signed-off-by: Emilio Benavente <emilio.benavente@nxp.com>`
Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
west.yml: Get changes for MCXN947 support

Get HAL updates for MCXN947 support

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

nordicjm commented on boards/nxp/frdm_mcxn947/board.cmake:
```diff
@@ -0,0 +1,18 @@
+#
+
+if(CONFIG_SOC_MCXN947_CPU0)
+board_runner_args(jlink "--device=MCXN947_M33_0" "--reset-after-load")
+board_runner_args(linkserver  "--device=MCXN947:MCX-N9XX-EVK:cm33_core0")
+board_runner_args(linkserver  "--override=/device/memory/1/flash-driver=MCXN9xx_S.cfx")
+board_runner_args(linkserver  "--override=/device/memory/1/location=0x10000000")
```
indent with 2x spaces for cmake

mmahadevan108 replied:
Thank you, done

nordicjm commented on soc/nxp/mcx/nx4x/Kconfig.soc:
```diff
@@ -0,0 +1,34 @@
+	select SOC_FAMILY_NXP_MCX
+
+config SOC_SERIES
+	default "nx4x" if SOC_SERIES_MCXNX4X
```
```suggestion
	default "mcxnx4x" if SOC_SERIES_MCXNX4X
```
And update soc.yml and board.yml

mmahadevan108 replied:
I have made this change

danieldegrasse commented on soc/nxp/mcx/mcxnx4x/linker.ld:
I think we can drop this and set the `SOC_LINKER_SCRIPT` variable directly to `zephyr/arch/arm/cortex_m/scripts/linker.ld` in the SOC CMakeLists file

nordicjm replied:
Agreed

danieldegrasse commented on soc/nxp/mcx/mcxnx4x/Kconfig.soc:
```diff
@@ -0,0 +1,34 @@
+	select SOC_MCXN947
+
+config SOC
+	default "mcxn947" if SOC_MCXN947_CPU0 || SOC_MCXN947_CPU1
```
can just be `if SOC_MCXN947`

danieldegrasse commented on soc/nxp/mcx/mcxnx4x/soc.c:
```diff
@@ -0,0 +1,81 @@
+#ifdef CONFIG_PLATFORM_SPECIFIC_INIT
+
+/* This function is taken from the SDK */
+static void system_init(void)
```
Can we call `SystemInit` directly here? We also may need to handle stack initialization here if we call any functions

mmahadevan108 committed:
doc: frdm_mcxn947: Add documentation for FRDM-MCXN947

Add documentation for FRDM-MCXN947

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

danieldegrasse commented on boards/nxp/frdm_mcxn947/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+==================
+
+The default configuration file
+``boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0_defconfig``
```
```suggestion
:zephyr_file:`boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0_defconfig`
```

mmahadevan108 replied:
Thanks, I have changed it.

nordicjm replied:
Not resolved

mmahadevan108 replied:
I had pushed the change. PR has been updated with this change.

danieldegrasse commented on boards/nxp/frdm_mcxn947/frdm_mcxn947_mcxn947_cpu0.yaml:
```diff
@@ -0,0 +1,19 @@
+  - gnuarmemb
+  - xtools
+supported:
+  - arduino_serial
```
Should also have `-gpio`

mmahadevan108 replied:
Thanks, I have added it.

mmahadevan108 commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1066,8 +1070,17 @@ static int mcux_lpuart_init(const struct device *dev)
+	/* When using LP Flexcomm driver, register the interrupt handler
+	 * so we receive notification from the LP Flexcomm interrupt handler.
+	 */
+	mcux_lpflexcomm_setirqhandler(config->parent_dev, dev,
```
The `parent_dev` would still be needed as the `nxp_lp_flexcomm_setirqhandler` function uses it to retrieve the pointer to `nxp_lp_flexcomm_data` structure where the ISR callbacks are stored.

decsny commented on boards/nxp/frdm_mcxn947/Kconfig:
```diff
@@ -0,0 +1,8 @@
+# Copyright 2024, NXP
```
remove comma

decsny commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-23, NXP
+ * Copyright (c) 2020-24, NXP
```
remove (c), remove comma, put full year

decsny commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -33,6 +33,28 @@ static int mcux_lpc_syscon_clock_control_on(const struct device *dev,
+	default:
+		break;
+	}
+#endif /* defined(CONFIG_PINCTRL_NXP_KINETIS) */
```
if we are putting here the ports and the flexcomms, can't we also put the GPIOs?

mmahadevan108 replied:
The GPIO driver needs to be updated for this which is beyond the scope of this PR and can be addressed as a separate PR.

decsny replied:
the gpio driver is already being updated significantly in this PR?

mmahadevan108 replied:
However adding clocking would involve updating other SoC's which use this driver and I do not want to address it in this PR.

decsny replied:
ok, fine, not a  big deal

decsny commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 2016 Freescale Semiconductor, Inc.
- * Copyright (c) 2017, NXP
+ * Copyright (c) 2017, 2024 NXP
```
remove (c) remove comma

decsny commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -96,7 +96,7 @@ static int gpio_mcux_configure(const struct device *dev,
 
 	} else if ((flags & GPIO_PULL_DOWN) != 0) {
-		/* Enable the pull and select the pulldown resistor (deselect
+		/* Enable the pull and select the pulldown resistor deselect
```
replace parentheses with comma instead, the grammar doesnt make sense here with no punctuation

decsny commented on drivers/pinctrl/pinctrl_kinetis.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2022-2023 NXP
+ * Copyright (c) 2022-2024 NXP
```
remove (c)

decsny commented on dts/arm/nxp/nxp_mcxn94x.dtsi:
```diff
@@ -0,0 +1,40 @@
+#include <mem.h>
+#include <arm/armv8-m.dtsi>
+#include <zephyr/dt-bindings/gpio/gpio.h>
+#include <zephyr/dt-bindings/i2c/i2c.h>
```
gpio and i2c headers not used in this file

decsny commented on dts/bindings/misc/nxp,lp-flexcomm.yaml:
```diff
@@ -0,0 +1,16 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+description: Lower Power Flexcomm node
```
```suggestion
description: Low Power Flexcomm
```

decsny commented on dts/bindings/pinctrl/nxp,kinetis-pinctrl.yaml:
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2022-2023, NXP
+# Copyright (c) 2022-2024, NXP
```
remove (c) remove comma

decsny commented on include/zephyr/dt-bindings/clock/mcux_lpc_syscon_clock.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020-2023, NXP
+ * Copyright 2020-2024, NXP
```
remove comma

decsny commented on modules/Kconfig.mcux:
```diff
@@ -1,12 +1,13 @@
 # MCUXpresso SDK
 
-# Copyright (c) 2016, Freescale Semiconductor, Inc.
+# Copyright 2016,2024 NXP
```
@yvanderv provide guidance

mmahadevan108 replied:
I will retain the Freescale Copyright and add NXP, this is after a conversation with @yvanderv

decsny commented on soc/nxp/mcx/mcxnx4x/soc.c:
```diff
@@ -0,0 +1,81 @@
+	/* SPI cannot be exist with UART or I2C on the same FlexComm Interface
+	 * Throw a build error if user is enabling SPI and UART/I2C on a Flexcomm node.
+	 */
+	DT_FOREACH_STATUS_OKAY(nxp_imx_lpspi, FLEXCOMM_CHECK)
```
this is not a runtime function, is a bit strange to put it here formatted as if it is, also, should this whole checking be done in a driver instead of soc.c?

mmahadevan108 replied:
I did not put this in the LPFlexomm driver as I felt the types of child nodes associated with `lpflexcomm` could change in a different soc.
Also we wanted to make this a build time failure so users are aware this combination of node enablement is not  supported by the hardware.

decsny replied:
this function is empty is my point, why is there a system init that does nothing

mmahadevan108 replied:
Nothing else needed for now besides this check. The clock init is done in the board folder.

decsny replied:
the check does not need to be within a function, it is a build time static assertion, it should be moved to avoid confusion

decsny commented on soc/nxp/mcx/mcxnx4x/soc.c:
```diff
@@ -0,0 +1,81 @@
+ * @brief Perform basic hardware initialization
+ *
+ * Initialize the interrupt controller device drivers.
+ * Also initialize the timer device driver, if required.
```
this  function doesn't appear to do anything, contrary to what this comment would suggest

mmahadevan108 replied:
Deleted the comment block. Leftover from a file copy.

decsny commented on soc/nxp/mcx/mcxnx4x/soc.h:
```diff
@@ -0,0 +1,27 @@
+#endif
+
+#ifdef __cplusplus
+}
```
there is nothing between these brackets

mmahadevan108 replied:
Preparing this for future work.

decsny replied:
doesn't hurt anything I guess

decsny reviewed (changes requested):
i have yet to review the lp flexcomm

nordicjm commented on boards/nxp/frdm_mcxn947/doc/frdm_mcxn947.jpg:
Smaller webp would be preferred

mmahadevan108 replied:
Sure, I will upload a webp file.

mmahadevan108 committed:
dts: nxp_lpuart: Interrupts is no longer a required property

Some NXP SoC's have a FlexComm interface which manages
interrupts.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017,2021 NXP
+ * Copyright (c) 2017,2021,2023,2024 NXP
```
2023 is needed after checking with our TA.

nordicjm reviewed (dismissed):
Build system parts OK

decsny commented on drivers/clock_control/clock_control_mcux_syscon.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020-23, NXP
+ * Copyright 2020-24 NXP
```
missed this before

mmahadevan108 replied:
This looks fine, am I missing something?

decsny replied:
it should be 2020-2024

decsny commented on drivers/gpio/gpio_mcux.c:
```diff
@@ -1,6 +1,6 @@
 /*
  * Copyright (c) 2016 Freescale Semiconductor, Inc.
- * Copyright (c) 2017, NXP
+ * Copyright 2017, 2023, 2024 NXP
```
should be 2023-2024

decsny commented on drivers/misc/CMakeLists.txt:
```diff
@@ -9,3 +9,4 @@ add_subdirectory_ifdef(CONFIG_TIMEAWARE_GPIO timeaware_gpio)
 add_subdirectory_ifdef(CONFIG_DEVMUX devmux)
 add_subdirectory_ifdef(CONFIG_NORDIC_VPR_LAUNCHER nordic_vpr_launcher)
 add_subdirectory_ifdef(CONFIG_MCUX_FLEXIO mcux_flexio)
+add_subdirectory_ifdef(CONFIG_NXP_LP_FLEXCOMM nxp_lp_flexcomm)
```
should go in mfd not misc

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017,2021 NXP
+ * Copyright 2017,2021,2023,2024 NXP
```
2023-2024

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -36,6 +39,7 @@ struct lpuart_dma_config {
 
 struct mcux_lpuart_config {
 	LPUART_Type *base;
+	const struct device *parent_dev;
```
just an observation, we could not build in this field if the CONFIG_NXP_LP_FLEXCOMM is not enabled, to save a few words of flash (just one for each flexcomm uart, so its not that important, ie not a blocking suggestion)

decsny commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -21,6 +21,9 @@
 
 #include <fsl_lpuart.h>
+#if CONFIG_NXP_LP_FLEXCOMM
+#include "nxp_lp_flexcomm.h"
```
im a bit confused how is this file found ?

mmahadevan108 replied:
Updated to move this driver to the `mfd` folder and the include file to `include/zephyr/drivers/mfd`

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.h:
I would say this file should be in include/ but it appears somehow the file is found by the other drivers using `#include "..."` ? would like to know how

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.c:
```diff
@@ -0,0 +1,169 @@
+#include <zephyr/logging/log.h>
+#include "nxp_lp_flexcomm.h"
+
+LOG_MODULE_REGISTER(nxp_lp_flexcomm, LOG_LEVEL_ERR);
```
should be configurable log level so that it can be turned off (ie LOG_LEVEL_NONE should be a possibility for this module)

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.c:
```diff
@@ -0,0 +1,169 @@
+
+struct nxp_lp_flexcomm_data {
+	struct nxp_lp_flexcomm_child *children;
+	size_t num_children;
```
seems that an array with a defined size can be described within this struct instead of adding a pointer and a size

mmahadevan108 replied:
I would prefer we keep the current implementation.

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.c:
```diff
@@ -0,0 +1,169 @@
+
+struct nxp_lp_flexcomm_child {
+	const struct device *dev;
+	uint8_t periph;
```
do  the dev and periph fields really need to go in the RAM? device_dt_get can get the child devices at compile time and periph seems it can be determined also at compile time unless I'm mistaken.

mmahadevan108 replied:
Looking at the driver design thats how this is needed to be.

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.c:
```diff
@@ -0,0 +1,169 @@
+struct nxp_lp_flexcomm_child {
+	const struct device *dev;
+	uint8_t periph;
+	child_isr_t lp_flexcomm_child_isr;
```
I will admit that registering the isr at runtime is something I did with the ENET driver since I also had to deal with a parent driver to handle IRQs, but I want to consider an approach of defining a linker section for a table of irq handlers for these parent drivers, if you dont want to deal with it in this PR then it's fine, but, combining this with the other comments I have on this struct, could remove all the RAM usage of this parent driver.

decsny commented on drivers/misc/nxp_lp_flexcomm/nxp_lp_flexcomm.c:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	/* If SPI is enabled with another interface type return an error */
+	if (spi_found && (i2c_found || uart_found)) {
```
a lot of this init function should be able to be determined at build time as well I think

mmahadevan108 replied:
I cannot think of a way to cover this logic at build time.

decsny reviewed (changes requested):
not all of the comments are blocking

blocking requested changes in this latest review:
- copyrights
- mfd instead of misc
- log level configurability

mmahadevan108 committed:
drivers: uart: Use the register define instead of enum define

The enum define is not implemented with the same name on all SoC's.
Use the register define instead.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

mmahadevan108 committed:
drivers: uart: Use the NXP Flexcomm driver to interrupt handling

The Low Power Flexcomm driver manages the interrupt handling
and provides an API to register interrupt callbacks.
Register the NXP LPUART interrupt handler.

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>
